pkg/operator/ceohelpers/qourum_check_test.go

```diff
@@ -96,13 +108,41 @@ func TestQuorumCheck_IsSafeToUpdateRevision(t *testing.T) {
 				u.WithNodeStatusAtCurrentRevision(3),
 				u.WithNodeStatusAtCurrentRevision(3),
 			),
+			nodeList: []*corev1.Node{
+				u.FakeNode("master-0", u.WithMasterLabel(), u.WithNodeInternalIP("10.0.0.1")),
+				u.FakeNode("master-1", u.WithMasterLabel(), u.WithNodeInternalIP("10.0.0.2")),
+				u.FakeNode("master-2", u.WithMasterLabel(), u.WithNodeInternalIP("10.0.0.3")),
+			},
 			etcdMembers: []*etcdserverpb.Member{
 				u.FakeEtcdMemberWithoutServer(0),
 				u.FakeEtcdMemberWithoutServer(1),
 			},
 			safe:        false,
 			expectedErr: fmt.Errorf("etcd cluster has quorum of 2 which is not fault tolerant: [{Member:name:\"etcd-0\" peerURLs:\"https://10.0.0.1:2380\" clientURLs:\"https://10.0.0.1:2907\"  Healthy:true Took: Error:<nil>} {Member:ID:1 name:\"etcd-1\" peerURLs:\"https://10.0.0.2:2380\" clientURLs:\"https://10.0.0.2:2907\"  Healthy:true Took: Error:<nil>}]"),
 		},
+		{
+			name: "Incomplete Quorum during removing one master",
+			objects: []runtime.Object{
+				u.BootstrapConfigMap(u.WithBootstrapStatus("complete")),
+			},
+			staticPodStatus: u.StaticPodOperatorStatus(
+				u.WithLatestRevision(3),
+				u.WithNodeStatusAtCurrentRevision(3),
+				u.WithNodeStatusAtCurrentRevision(3),
+				u.WithNodeStatusAtCurrentRevision(3),
+			),
+			nodeList: []*corev1.Node{
+				u.FakeNode("master-0", u.WithMasterLabel(), u.WithNodeInternalIP("10.0.0.1")),
+				u.FakeNode("master-1", u.WithMasterLabel(), u.WithNodeInternalIP("10.0.0.2")),
+			},
+			etcdMembers: []*etcdserverpb.Member{
+				u.FakeEtcdMemberWithoutServer(0),
+				u.FakeEtcdMemberWithoutServer(1),
+				u.FakeEtcdMemberWithoutServer(2),
+			},
+			safe:        false,
+			expectedErr: fmt.Errorf("CheckSafeToScaleCluster 3 nodes are required, but only 2 are available"),
+		},
```

Comment:
should we fall into this situation as the UT I added ?: Etcd cluster is healthy, all nodes are ready(Just change one node label from master to worker ) @tjungblu

Reply:
>  all nodes are ready(Just change one node label from master to worker 

that's an interesting case, how would the static pod react to that?

Reply:
- if we only remove the master role: node-role.kubernetes.io/master, the static pods on that node could still be running as normal, the etcd cluster is healthy too,:
```
https://x.x.x.1:2379, 5df6bd632e46a20e, 3.5.11, 127 MB, false, false, 10, 2458929, 2458929,
https://x.x.x.2:2379, 1b6b8161626a4eb9, 3.5.11, 127 MB, true, false, 10, 2458929, 2458929,
https://x.x.x.3:2379, 49869657b1942ea7, 3.5.11, 128 MB, false, false, 10, 2458929, 2458929,
```

but the ceo logs show as follows because there are only two masters now:

```
I0411 01:46:23.347054       1 quorumguardcleanupcontroller.go:133] 3/2 guard pods ready. Waiting until all new guard pods are ready
I0411 01:46:23.399103       1 quorumguardcleanupcontroller.go:133] 3/2 guard pods ready. Waiting until all new guard pods are ready
I0411 01:46:23.447863       1 quorumguardcleanupcontroller.go:133] 3/2 guard pods ready. Waiting until all new guard pods are ready
I0411 01:46:23.498217       1 quorumguardcleanupcontroller.go:133] 3/2 guard pods ready. Waiting until all new guard pods are ready
I0411 01:46:23.550646       1 quorumguardcleanupcontroller.go:133] 3/2 guard pods ready. Waiting until all new guard pods are ready
I0411 01:46:24.448640       1 quorumguardcleanupcontroller.go:133] 3/2 guard pods ready. Waiting until all new guard pods are ready
E0411 01:46:25.598762       1 base_controller.go:272] TargetConfigController reconciliation failed: TargetConfigController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:25.646253       1 base_controller.go:270] "TargetConfigController" controller failed to sync "", err: TargetConfigController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:25.649209       1 base_controller.go:272] EtcdEndpointsController reconciliation failed: EtcdEndpointsController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:25.845918       1 base_controller.go:272] EtcdCertSignerController reconciliation failed: EtcdCertSignerController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:41.373595       1 base_controller.go:272] EtcdCertSignerController reconciliation failed: EtcdCertSignerController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:41.374949       1 base_controller.go:272] TargetConfigController reconciliation failed: TargetConfigController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:41.378822       1 base_controller.go:272] EtcdEndpointsController reconciliation failed: EtcdEndpointsController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:41.381235       1 base_controller.go:270] "TargetConfigController" controller failed to sync "", err: TargetConfigController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available
E0411 01:46:51.639452       1 base_controller.go:272] EtcdCertSignerController reconciliation failed: EtcdCertSignerController can't evaluate whether quorum is safe: CheckSafeToScaleCluster 3 nodes are required, but only 2 are available

```

Reply:
will one etcd pod eventually wind down? I didn't exactly found the scale-down logic here, but it's also based on the node status from:
https://github.com/openshift/library-go/blob/master/pkg/operator/staticpod/controller/node/node_controller.go#L47-L55

If it eventually shuts down that pod, failing the sync loops is the right thing to do.

Reply:
I just scale-down the cluster manually by relabeling the master node to worker, and remove the etcd  static pod yaml to simulate this situation in ocp 4.12